### ISSUE TITLE
Import by declarative code over imperative command

### ISSUE THREAD

**Issue by entropitor** [hashicorp/terraform#26364]
### Current Terraform Version
```
Terraform v0.12.29

Your version of Terraform is out of date! The latest version
is 0.13.3. You can update by downloading from https://www.terraform.io/downloads.html
```

### Use-cases
* Importing resources from multiple branches without deploying 1 branch deleting things imported by the other branch
* Previewing diffs without *actually* importing resources

### Attempted Solutions
* Running terraform import and coordinating the deploys
* Running terraform import, terraform plan and then terraform state remove

### Proposal
Have a declarative import-id meta argument that allows importing resources by id (if it doesn't exist in the state yet), just as importing by terraform command (see e.g. https://www.pulumi.com/docs/guides/adopting/import/)

* That way, we can "import" stuff across different branches without risking deleting some of those imported resources when not being careful enough with the plan (and it not blocking the deploy until the other PR that imported stuff is merged)
* That way, you can see a diff against an existing resources before deploying it

### References
_None_

**Comment by danieldreier** [comment]
Hi @entropitor! I think the general idea of managing changes via declarative code rather than imperative commands is a great idea, and it's consistent with some of our long-term thinking for how to evolve Terraform. I can't promise we'll do exactly this, but it's very helpful to hear someone in the community voice a similar sentiment to what we've been imagining internally. Thank you!

**Comment by entropitor** [comment]
I'm not 100% sure but I think something like this would have saved Spotify from deleting one of their clusters (https://datacamp.slack.com/archives/D7BUK8CV9/p1599206679007800), or made it less likely at least

**Comment by apparentlymart** [comment]
This use-case is in the same family as what's discussed in #9048.

I'm not going to close it because that issue is currently phrased specifically as being about `terraform state mv`, rather than generically for all of the possible side-effects, but [my recent comment there](https://github.com/hashicorp/terraform/issues/9048#issuecomment-700360009) is also relevant to a configuration-driven import mechanism, because the plan model redesign I mentioned there would also be a prerequisite for a "plannable" import.

**Comment by mwarkentin** [comment]
I've come across https://github.com/minamijoyo/tfmigrate

Haven't used it personally but maybe helpful for some people in this thread.

**Comment by awwright** [comment]
Someone on another issue I filed pointed me here and I'm having trouble figuring out what this issue is proposing, exactly. Can you please provide an example of a problem this solves, and how it solves it?

**Comment by bondsbw** [comment]
In #27959 I proposed that the existing `id` attribute (or equivalent, e.g. `arn`) be used as the import argument.  I feel this would increase clarity as they have the same definition and use the same value.

**Comment by bondsbw** [comment]
Additionally, in #27959 I added that the identifier may be supplied with variable values and interpolation.  This can open up more complex scenarios like effectively importing by tag, or importing based on the resolved values of dependencies:

```tf
# retrieve existing instance if it exists
data "aws_instance" "existing_server" {
  filter {
    name   = "tag:Name"
    values = ["instance-name-tag"]
  }
}

# create new instance if it does not exist
resource "aws_instance" "web" {
    id = try(data.aws_instance.existing_server.id, null)
    ...
}

# simultaneously import all security groups assigned to the instance
resource "aws_security_group" "web_security_groups" {
    for_each = aws_instance.web.security_groups
    id       = each.value
    ...
}
```

**Comment by rileykarson** [comment]
Some of the ideas discussed here are close to https://github.com/hashicorp/terraform/issues/19017, namely "you can see a diff against an existing resources before deploying it"

**Comment by reegnz** [comment]
Now that `moved` was done in the context of #19354, I think it would be great if we'd get something similar for this one, like `imported`:

```hcl
imported {
  id = "MyRole"
  to = aws_iam_role.myrole
}
```

**Comment by reegnz** [comment]
Related discussion over here: https://discuss.hashicorp.com/t/request-for-feedback-config-driven-refactoring/30730/5

**Comment by LDVSOFT** [comment]
I agree with all comments above, I just want to add my experience.

In documentation, `moved` is presented mostly as a tool for module upgrades where new module version uses new name for already existing resources, and so instead of recreating stacks it helps to move resources to their new names instead, assisting module users. I, however, mostly write leaf Terraform code and use `moved` blocks for refactorings: after I rename a resource of a module, I write a small `moved` block, apply the state and toss away `moved` (it's often isn't even commited, but for shared workflow it probably should have been). Moved blocks really help checking diffs and seeing that your refactor only actually renames resources and changes one or two instead of full destruction and re-creation (that actually can lose data and/or fail since there is a race for explicit IDs).

So, for me `moved` blocks is not as much a compatibility tool as a developer tool. Also, sometimes I write code that takes control over existing resources, and while doing this I often leave comments for future `import`s that will definitely happen:
```hcl
// TODO: import it! id='abacabadabacaba'
resource "provider_resource_a" "foo" {
```
just like `moved` above. If I really wanted I could've used a tool that runs `import` automatically based on those.

So, from my point of view, native declarative auto-imports would be a very valuable addition.

**Comment by awwright** [comment]
Hi, I'm still not sure what this issue is asking. I had my issue #27366 closed and directed to this one. My issue was specifically about computing the resource ID of a specific resource. And I'm not sure how this issue satisfies that feature request, if at all.

**Comment by reegnz** [comment]
@awwright Sometimes you cannot infer the resource id, as it's random, and generated on the server side when you first create the resource. None of the attributes determine what the ID might be, so you cannot deterministically determine what the ID will be. So you need to work backwards, and give terraform the id so it can perform the import. Because of that, we need an **explicit** (non-inferred) **declarative** import statement, that can perform the import in an **idempotent** manner.

**Comment by bondsbw** [comment]
@reegnz For clarification, by “non-inferred” we are not saying there is a restriction against deferred or interpolated values?

**Comment by reegnz** [comment]
I'm saying that the import ID cannot be determined by just looking at the attributes of the resource. There are some 'accidental' cases where one of the attributes is used as the id or part of the id, but that's not part of terraform's contract, terraform is oblivious to that happenstance.

Just a concrete example, you cannot infer an `aws_vpc` id just by looking at the resource attributes.
But for a counterexample, with `aws_s3_bucket` the id matches the bucket name.

The common case for both though, is that you don't infer the id from the attributes, but have it be explicitly defined.

IMHO the simplest case would be to only support explicit import id-s, you don't need inference. If you do, you can write something that parses the terraform and generates the import statements.

I do find it important to have an **explicit** block for imports though, so we remain in the `terraform apply` semantics, and we don't want resources auto-importing (that would break existing assumptions), unless there's an explicit import block.

**Comment by apparentlymart** [comment]
At the moment this issue is representing only the use-case of handling "importing" (that is: creating a binding to a pre-existing object in the Terraform state, instead of creating a new object) as part of a normal plan and apply run, using only information in the configuration.

Exactly how that might be achieved and what new language features it might involve still remains to be seen once there is a full design effort for this. For the moment, any solution that meets the need is on the table for consideration.

**Comment by bondsbw** [comment]
@reegnz I don’t think an explicit import action is a necessary restriction.

Example:  Say for a particular resource, `id` is chosen as the variable you set when you want to import. Then the cases seem unambiguous:

1. No value is set for `id`:  follow existing behavior (no import)
2. A value is set, and a resource with that ID does not exist at the provider:  error (bad ID; cannot import)*
3. A value is set, a resource with that ID exists at the provider, and the resource exists in state with the same ID: no change
4. A value is set, a resource with that ID exists at the provider, but an old resource exists in state with a different ID which also exists at the provider:  plan to delete the old resource and import the new one
5. A value is set, a resource with that ID exists at the provider, but an old resource exists in state with a different ID but does not exist at the provider:  plan only to import the new one

* Case 2 assumes the ID has a resolved value.  If the value for `id` is not yet known, then this logic will be carried out only once it is resolved.

@apparentlymart I’m concerned that such a narrow definition might encourage a design that excludes information from data sources or other resources, and perhaps makes it impractical to extend to such use cases in the future.  I’d like that to be a goal even if not implemented at first.

**Comment by awwright** [comment]
> I'm saying that the import ID cannot be determined by just looking at the attributes of the resource.

I address this in [my issue](https://github.com/hashicorp/terraform/issues/27366); the import ID is not _necessarily_ determined by the known attributes of the resource, but it is nonetheless sometimes possible. Even if the ID is based on random, server-generated data, there's nonetheless cases where there's only one ID that would be valid to import with. Please look at the example where I use random_string to generate a bucket name.

Since this is a straightforward function (Can I determine the resource ID? If so, what is it? If not, what information am I missing?), and since it would apply to many issues similar to this, I think this part should be broken off into a separate issue.

**Comment by reegnz** [comment]
Still, I think infering an import ID is an extension of the initial feature request, which is to import resources where you cannot determine the ID. If you can infer the ID in some cases, that's great. But for the common-case, you cannot assume that.
I think that infering should be out-of-scope for this issue, and let's just make the basic case work: import when user provides the ID.

**Comment by apparentlymart** [comment]
A `data` block does potentially qualify as information in the configuration. My intent with that qualification was to exclude anything involving special CLI commands or arguments that would be relevant only during importing, and that would potentially require a user of a module to take some special action after upgrading that module.

There are several design questions here around who gets to determine which specific objects are to be imported and how they would be specified. Something involving data sources would be on the table in principle. The design space here is wide open at the moment.

**Comment by ajf-firstup** [comment]
Has there been any more discussion of this feature? I'm currently trying to bring a large and somewhat unruly production setup under Terraform control, and having some level of "auto import" would be fantastic.

I have some ideas on possible syntaxes and lifecycle issues, but I have relatively limited exposure to TF (only a year or two, and almost entirely on AWS) so seeing what the gurus have come up with / discussed / discarded would be very enlightening.

**Comment by awwright** [comment]
@kmoe I'm not sure this is the same feature as in #31455. If it is, can you please update the description in this issue? The title and description of this issue is completely incomprehensible.

**Comment by apparentlymart** [comment]
I'm not @kmoe but I think the idea here is that #31455 is proposing one specific design for importing (based on data sources) whereas this issue is representing the problem space in general, without limiting to any particular design.

We tend to use issues to represent problems or new capabilities rather than directly represent proposed solutions, so typically we try to consolidate issues that propose specific solutions to problems together into a single discussion about the problem space so that we can more easily see all of the discussion and proposed solutions in one place. Because that closed issue now refers to this one we can see the link to it in the sequence above and so still consider it while evaluating a variety of possible designs to meet this request.

Thanks!

**Comment by awwright** [comment]
@apparentlymart Except this issue wasn't written to talk about "the problem space in general" either... just re-read the description

> Importing resources from multiple branches without deploying 1 branch deleting things imported by the other branch

> Previewing diffs without actually importing resources

This issue has _nothing_ in common with #31455, or my issue #27366, or any of the other much better phrased issues than this one.

Maybe you're talking about the "proposal" heading that _seems_ to talk about importing resources by their ID, but that's no solution for my issue at all, but this didn't stop you guys from closing it.

**Comment by apparentlymart** [comment]
Hi @awwright,

I'm not really sure what you are proposing that we do in response to these observations. Are you suggesting that a different issue should be the one that we choose to consolidate into, rather than this one?

When I refer to a problem or use-case I'm talking about the part of the original description under the "Use-case" heading. Although the "Proposal" there is one of the proposal in consideration, it is not what this issue actually represents in our process. If there's another issue that you think better captures the _problem statement_ then we could consider how to represent that, but so far this was just the issue that we noticed first or the one that attracted the most discussion already and so is the one we've arbitrarily chosen as the placeholder for this problem.

That doesn't imply that the original proposal in this has any more weight than any other proposal, because at this stage there is _no_ specific chosen approach. All viable approaches are on the table to be evaluated when the time comes to prioritize these use-cases. At that time we will still refer to all of the relevant issues, including the ones that are closed. The goal here is only to ensure there is a single place to look to find all of the links to all of the relevant proposals and discussion.

**Comment by awwright** [comment]
@apparentlymart I'm proposing you rewrite the title and the description. Consolidating into a different issue that better explains the problem would also suffice.

---

My complaint with the process you outlined is there are multiple different problems getting merged together that no one solution will solve.

My issue #27366 is fundamentally quite narrow: Given a document like

```
resource "aws_s3_bucket" "example" {
   bucket = "my_bucket_example"
}
```

... This document contains all the information necessary to compute its address. Given the ID of this resource, we also know its address. It doesn't involve changes to defining resources. There are no new arguments. It doesn't have anything to do with branches. It has nothing in common with this issue; you could implement and close #27366 long before you ever get started on this one.

**Comment by ervin-pactum** [comment]
I would like to add, that without this, imports on terraform cloud seem to be impossible, and some of admins has to configure environment locally to do import(s), including setup of some of secrets and most of providers.

**Comment by treksler** [comment]
i just discovered the moved block (i know, where have i been?)
so i naturally assumed there would also be an import block, because why would you implement one, without doing the other? the problem that is being solved is the exact same. move imperative actions to declarative code

turns out that is not so

the obvious solution is having import blocks
kind of like
https://github.com/hashicorp/terraform/issues/26364#issuecomment-995024469


i would suggest matching the declarative syntax to the existing imperative syntax

```
The import command expects two arguments.
Usage: terraform [global options] import [options] ADDR ID
```

so this is what i think should be implemented
!!!!

```
import {
  addr = aws_iam_role.myrole
  id = "MyRole"
}
```
!!!!

this could be implemented today  if we wanted to. 
there is zero new logic, it would simply do exactly what `terraform import` already does.
boiling the ocean can wait for another time

if the resource already exists, there should be no failure, because otherwise it would fail on every subsequent run.

a good use case for this is any aws resources that aws creates automatically on first run.
I can't really use data blocks to read them because i have no way of knowing if they have been created already by someone else. so i am stuck creating TF resources, BUT if the resource already exists, then i need to import them first. big PITA. i would like to be able to just write the code, deploy it and have it work without having to do mental gymnastics about what imperative commands i have to run for which terraform states

adding an import block to the code would solve the issue (import if already exists, otherwise create)

having this available would also open up the possibility of someone writing a tfautoimport tool whose job would be to analyse a terraform plan, and to connect to a provider (eg aws) and analyse what already exists there and to write out some suggested import blocks (kind of like tfautomv does for moved blocks)

**Comment by lorengordon** [comment]
@treksler agree 1000%. i don't need any magic auto-import functionality. just a declarative block that let's me define the import as config so plan can expect it and apply can execute it. i'll do the work to write the the config.

> i just discovered the moved block (i know, where have i been?) so i naturally assumed there would also be an import block, because why would you implement one, without doing the other? the problem that is being solved is the exact same. move imperative actions to declarative code
> 
> turns out that is not so
> 
> the obvious solution is having import blocks kind of like [#26364 (comment)](https://github.com/hashicorp/terraform/issues/26364#issuecomment-995024469)
> 
> i would suggest matching the declarative syntax to the existing imperative syntax
> 
> ```
> The import command expects two arguments.
> Usage: terraform [global options] import [options] ADDR ID
> ```
> 
> so this is what i think should be implemented !!!!
> 
> ```
> import {
>   addr = aws_iam_role.myrole
>   id = "MyRole"
> }
> ```
> 
> !!!!
> 
> this could be implemented today if we wanted to. there is zero new logic, it would simply do exactly what `terraform import` already does. boiling the ocean can wait for another time
> 
> if the resource already exists, there should be no failure, because otherwise it would fail on every subsequent run.
> 
> a good use case for this is any aws resources that aws creates automatically on first run. I can't really use data blocks to read them because i have no way of knowing if they have been created already by someone else. so i am stuck creating TF resources, BUT if the resource already exists, then i need to import them first. big PITA. i would like to be able to just write the code, deploy it and have it work without having to do mental gymnastics about what imperative commands i have to run for which terraform states
> 
> adding an import block to the code would solve the issue (import if already exists, otherwise create)
> 
> having this available would also open up the possibility of someone writing a tfautoimport tool whose job would be to analyse a terraform plan, and to connect to a provider (eg aws) and analyse what already exists there and to write out some suggested import blocks (kind of like tfautomv does for moved blocks)

**Comment by tomwidmer** [comment]
Is it useful to be able to mark the import block as required or not? e.g.
```
import {
  addr = aws_iam_role.myrole
  id = "MyRole"
  required = true
}
```
meaning that, if the resource doesn't exist in tf state, and the id doesn't exist, it's an error. Defaults to `false`, meaning it's more permissive, and if both the resource and the id don't exist, TF just creates the resource as usual.

This would make it so much easier to do refactorings such as importing and managing resources 'created behind your back' like AWS CloudWatch log groups, etc.

**Comment by geekofalltrades** [comment]
I see that this is coming up in Terraform 1.5.0. Has documentation been added for it? Where can I find it?

**Comment by camaeel** [comment]
PR https://github.com/hashicorp/terraform/pull/31910 is still in draft.

**Comment by crw** [comment]
@camaeel please see https://github.com/hashicorp/terraform/pull/33085.

**Comment by crw** [comment]
@geekofalltrades Thanks for the suggestion! Docs are in-progress, as the release is forthcoming it should not be too much longer before they are available for review.

**Comment by lorengordon** [comment]
Just curious what is keeping this issue open still, now that 1.5 has been released with support for `import` blocks?

**Comment by kmoe** [comment]
Quite right, this is done with `import` blocks.

**Comment by awwright** [comment]
Would someone please re-open my issue #27366? It was merged into this one, but import blocks don't solve the problem I identified in that issue.
